lesson_2/mortgage_calc.py

```diff
@@ -0,0 +1,158 @@
+''' build a mortgage calculator that takes
+in loan amount, APR, and loan duration and returns the
+monthly payment assuming that interest is compounded monthly.'''
+
+# TODO: break up loan_calculator to be 10-15 lines only, make more functions of the small moving parts
+# TODO: make sure function names reference their side effect (ex display_intro_outro)
+import json
+with open('mortgage_messages.json', 'r') as file:
+    MESSAGES = json.load(file)
+
+def prompt(message):
+    print(f"==> {message}")
+
+def intro_outro(message):
+    print(f"-------{message}-------")
+    print()
+
+def error_output(message):
+    print(f"\n    !!!! {message} !!!!\n\n    -------TRY AGAIN!-------")
+
+def get_loan_amount():
+    prompt(MESSAGES["loan_amount"])
+    loan_amount_input = input()
+    return loan_amount_input
+
+def invalid_loan_amount(input_value):
+    try:
+        float(input_value)
+        if float(input_value) <= 0:
+            return True
+    except ValueError:
+        return True
+    return False
+
+def get_apr():
+    prompt(MESSAGES["APR"])
+    annual_percentage_rate_input = input()
+    return annual_percentage_rate_input
+
+def invalid_apr(input_apr):
+    try:
+        float(input_apr)
+        if float(input_apr) < 0:
+            return True
+    except ValueError:
+        return True
+    return False
```

Comment:
`invalid_loan_ammount` and `invalid_apr` are doing the same thing, the only difference is the name of their parameters...

For example, `invalid_loan_ammount` and `invalid_apr` are really just checking if the input is a float and returning `true` when it is `<0`, `false` otherwise. This can be a single function:

```py
def invalid_float(input_value):
    try:
        input_as_a_float = float(input_value)
        if input_as_a_float <= 0:
            return True
    except ValueError:
        return True
    return False
```

Something really important to note here is that I don't use `float(input_apr)` twice. Instead of doing the type conversion twice, it's more efficient to store it in a variable (`input_as_a_float`) and reuse it in the `if` block.

Reply:
ahh I see